### PR TITLE
Ensure WeakOrderQueue can be collected fast enough

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

Commit afafadd3d7caf1e4b346da049baab0afeae0a4bc introduced a change which stored the Stack in the WeakOrderQueue as field. This unfortunally had the effect that it was note removed from the WeakHashMap anymore as the Stack also is used as key.

Modifications:

Not store a reference to the Stack in WeakOrderQueue.

Result:

WeakOrderQueue can be collected correctly again.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@vietj can you check ?

**Comment by normanmaurer** [comment]
This was first reported in the netty group:
https://groups.google.com/forum/#!msg/netty/wJ-fzhsSMx8/aEXf8z5SBAAJ

**Comment by normanmaurer** [comment]
@Scottmitch FYI

**Comment by vietj** [comment]
@alexlehm is going to check the behavior in our CI

**Comment by Scottmitch** [comment]
`it was note removed` -> `it was not removed`

**Comment by Scottmitch** [comment]
`Not store a reference` -> `Do not store a reference`

File: common/src/main/java/io/netty/util/Recycler.java [hunk 1]
```diff
@@ -232,19 +232,42 @@ public void recycle(Object object) {
                 next = stack.head;
                 stack.head = this;
             }
-            this.stack = stack;
+
+            // Its important that we not store the Stack itself in the WeakOrderQueue as the Stack also is used in
+            // the WeakHashMap as key. So just store the enclosed AtomicInteger which should allow to have the
+            // Stack itself GCed.
+            availableSharedCapacity = stack.availableSharedCapacity;
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/Recycler.java]: the stack doesn't actually use this variable ... should we move it outside the stack and lazy create it here in the `WeakOrderQueue`?

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/Recycler.java]: @Scottmitch nope as it needs to be shared across all `WeakOrderQueue`instances  that belong to a `Stack`.

Comment (Scottmitch) [inline common/src/main/java/io/netty/util/Recycler.java]: +1 makes sense

**Comment by normanmaurer** [comment]
@vietj @alexlehm cool. Let me know once you know more...

**Comment by vietj** [comment]
@normanmaurer it looks like it's positive but Alex will confirm

**Comment by alexlehm** [comment]
@normanmaurer the pr fixes the issue we are seeing in the vert.x tests, with the recycler_weak_hash_map branch, it works correctly

**Comment by vietj** [comment]
awesome ^^

**Comment by normanmaurer** [comment]
@vietj @alexlehm WOOOOT... @Scottmitch let me know if you are done with review and I will pull in.

**Comment by Scottmitch** [comment]
ship it ... good work!

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (94d7557dead5e7780769d44bbb4aa7530149ca34) and 4.0 (688e65f)

**Pull request closed without merging**